tests/unit/test_charm.py

```diff
@@ -439,150 +440,101 @@ def test_unit_ips(self):
-                        continue
-
-                    self.assertIn(
-                        "Replica set for units: ['1.1.1.1'], is not ready", "".join(logs.output)
-                    )
-                    self.assertEqual(
-                        self.harness.charm.unit.status,
-                        BlockedStatus("failed to initialise replica set"),
-                    )
+        reconfigure.assert_not_called()
 
     @patch_network_get(private_address="1.1.1.1")
-    @patch("charm.MongodbOperatorCharm._initialise_replica_set")
-    @patch("mongod_helpers.MongoDB.is_ready")
-    @patch("charm.MongodbOperatorCharm.app")
-    def test_mongodb_relation_joined_peers_not_ready(self, app, is_ready, initialise_replica_set):
+    @patch("mongod_helpers.MongoDB.reconfigure_replica_set")
+    @patch("charm.MongodbOperatorCharm._single_mongo_replica")
+    @patch("charm.MongodbOperatorCharm._peers")
+    def test_mongodb_relation_joined_peers_not_ready(self, _, single_replica, mongo_reconfigure):
         # preset values
         self.harness.set_leader(True)
-        is_ready.return_value = False
+        single_replica.return_value.is_ready.return_value = False
 
         # simulate 2nd MongoDB unit
         rel = self.harness.charm.model.get_relation("mongodb")
         rel_id = rel.id
         key_values = {"private-address": "127.4.5.6"}
         self.harness.add_relation_unit(rel_id, "mongodb/1")
         self.harness.update_relation_data(rel_id, "mongodb/1", key_values)
-        app.planned_units.return_value = len(self.harness.charm._unit_ips)
 
-        with self.assertLogs("charm", "DEBUG") as logs:
-            self.harness.charm.on.mongodb_relation_joined.emit(relation=rel)
+        # verify that we do not reconfigure replica set
+        mongo_reconfigure.assert_not_called()
 
-            self.assertIn(
-                "unit: <ops.model.Unit mongodb/1> is not ready, cannot "
-                + "initialise replica set until all units are ready, deferring "
-                + "on relation-joined",
-                "".join(logs.output),
-            )
+    @patch_network_get(private_address="1.1.1.1")
+    @patch("charm.MongodbOperatorCharm._single_mongo_replica")
+    @patch("charm.MongodbOperatorCharm._need_replica_set_reconfiguration")
+    @patch("mongod_helpers.MongoDB.reconfigure_replica_set")
+    def test_mongodb_relation_joined_peer_ready(
+        self, mongodb_reconfigure, need_reconfiguration, single_replica
+    ):
+        # preset values
+        self.harness.set_leader(True)
+        single_replica.return_value.is_ready.return_value = True
+
+        # test cases where replica set should and shouldn't be reconfigured
+        for reconfiguration in [False, True]:
+            self.harness.charm._need_replica_set_reconfiguration = reconfiguration
+
+            # simulate 2nd MongoDB unit
+            rel = self.harness.charm.model.get_relation("mongodb")
+            rel_id = rel.id
```

Comment:
Hi @zeljkoagic I haven't had my morning coffee yet, so I am a little confused, what part seems redundant? :) 

I do 
```
            # simulate 2nd MongoDB unit
            rel = self.harness.charm.model.get_relation("mongodb")
            rel_id = rel.id
            key_values = {"private-address": "127.4.5.6"}
            self.harness.add_relation_unit(rel_id, "mongodb/1")
            self.harness.update_relation_data(rel_id, "mongodb/1", key_values)
```
to triggered the `on_relation_joined_event` and `on_relation_changed_event`, it is important that we trigger them in this way so that when we call our hooks there is populated event data.

Reply:
Seems I caught multiple lines here, apologies. The only thing that's redundant here is minor: `rel_id = rel.id`. Why not just use `rel.id`?

Reply:
I see! Okay yes, agreed!